compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -0,0 +1,931 @@
+    return;
+  }
+
+  // Otherwise, try resolving with init.
+  ChangeResult changed = result->resolve(init);
+  update(result, changed);
+}
+
+static void propagateLayoutToTransposeOp(
+    vector::TransposeOp transpose,
+    ArrayRef<const DistributionLayout *> operandLattices,
+    ArrayRef<DistributionLayout *> resultLattices,
+    std::function<void(DistributionLayout *, ChangeResult)> update) {
+  // Transpose has only one vector result.
+  DistributionLayout *result = resultLattices[0];
+  // Transpose has only one vector operand.
+  const DistributionLayout *value = operandLattices[0];
+
+  // If result lattice already has a layout, we cannot do anything. We do not
+  // impose layout conflicts on results.
+  if (result->hasLayout()) {
+    return;
+  }
+
+  // Cannot propagate layout if value is uninitialized.
+  if (value->isUninitialized()) {
+    return;
+  }
+
+  // Build a transposed layout.
+  SmallVector<unsigned> permutation;
+  ArrayRef<int64_t> perm = transpose.getPermutation();
+  VectorLayoutInterface permutedLayout = value->getInnerLayout().permute(perm);
+
+  // Try to resolve with the transposed layout.
+  ChangeResult changed = result->resolve(permutedLayout);
+  update(result, changed);
+}
+
+void propagationTransferFunction(
+    Operation *op, ArrayRef<const DistributionLayout *> operandLattices,
+    ArrayRef<DistributionLayout *> resultLattices,
+    std::function<void(DistributionLayout *, ChangeResult)> update) {
+
+  // Propagate layout to elementwise operations.
+  if (OpTrait::hasElementwiseMappableTraits(op)) {
+    propagateLayoutToElementwiseOp(op, operandLattices, resultLattices, update);
+    return;
+  }
+
+  if (auto multiReduce = dyn_cast<vector::MultiDimReductionOp>(op)) {
+    propagateLayoutToMultiReductionOp(multiReduce, operandLattices,
+                                      resultLattices, update);
+    return;
+  }
+
+  if (auto transpose = dyn_cast<vector::TransposeOp>(op)) {
+    propagateLayoutToTransposeOp(transpose, operandLattices, resultLattices,
+                                 update);
+    return;
+  }
+
+  return;
+}
+
+/// ==========================================================================
+///        ENFORCEMENT TRANSFER FUNCTIONS
+/// ==========================================================================
+
+static void enforceLayoutToElementwiseOp(
+    Operation *op, ArrayRef<DistributionLayout *> operandLattices,
+    ArrayRef<const DistributionLayout *> resultLattices,
+    std::function<void(DistributionLayout *, ChangeResult)> update) {
+  // All operands and results must agree on the same layout.
+
+  // We do not support multiple results yet.
+  if (resultLattices.size() != 1)
+    return;
+
+  // Try to enforce the layout of the result on operands.
+  const DistributionLayout *result = resultLattices[0];
+  if (result->hasLayout()) {
+    // Note that the operand lattice is not updated. So using the operand
+    // lattice again can cause bugs.
```

Comment:
So seems like you first run the forward pass and during the backward pass add conflict resolution. How do we know that's better than doing the backward first and then the forward? Could we do forward and backward and maintain a list of possible layouts and then decide which is best based on minimizing layout conflicts?